webtau-graphql/src/main/java/org/testgisdocumenting/webtau/graphql/GraphQLSchema.java

```diff
@@ -0,0 +1,113 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.testgisdocumenting.webtau.graphql;
+
+import graphql.ExecutionInput;
+import graphql.ParseAndValidate;
+import graphql.ParseAndValidateResult;
+import graphql.language.OperationDefinition;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.idl.RuntimeWiring;
+import graphql.schema.idl.SchemaGenerator;
+import graphql.schema.idl.SchemaParser;
+import graphql.schema.idl.TypeDefinitionRegistry;
+import org.testingisdocumenting.webtau.http.request.HttpRequestBody;
+import org.testingisdocumenting.webtau.utils.FileUtils;
+import org.testingisdocumenting.webtau.utils.JsonUtils;
+
+import java.net.URI;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.stream.Stream;
+
+public class GraphQLSchema {
+    private final boolean isSchemaDefined;
+    private final Set<GraphQLOperation> schemaDeclaredOperations;
+
+    public GraphQLSchema(String schemaUrl) {
+        isSchemaDefined = !schemaUrl.isEmpty();
+        schemaDeclaredOperations = new HashSet<>();
+
+        if (isSchemaDefined) {
+            SchemaParser schemaParser = new SchemaParser();
+            Path schemaPath = Paths.get(URI.create(schemaUrl));
+            TypeDefinitionRegistry typeDefinitionRegistry = schemaParser.parse(FileUtils.fileTextContent(schemaPath));
+
+            RuntimeWiring runtimeWiring = RuntimeWiring.newRuntimeWiring().build();
+
+            SchemaGenerator schemaGenerator = new SchemaGenerator();
+            graphql.schema.GraphQLSchema schema = schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring);
+
+            BiConsumer<Optional<GraphQLObjectType>, GraphQLOperationType> registerTypes = (objectType, operationType) ->
+                    objectType.ifPresent(type ->
+                            type.getFieldDefinitions().forEach(definition -> schemaDeclaredOperations.add(
+                                    new GraphQLOperation(definition.getName(), operationType)
+                            )));
+            registerTypes.accept(Optional.ofNullable(schema.getQueryType()), GraphQLOperationType.QUERY);
+            registerTypes.accept(Optional.ofNullable(schema.getMutationType()), GraphQLOperationType.MUTATION);
+            registerTypes.accept(Optional.ofNullable(schema.getSubscriptionType()), GraphQLOperationType.SUBSCRIPTION);
+        }
+    }
+
+    public boolean isSchemaDefined() {
+        return isSchemaDefined;
+    }
+
+    public Stream<GraphQLOperation> getSchemaDeclaredOperations() {
+        return schemaDeclaredOperations.stream();
+    }
+
+    public Optional<GraphQLOperation> findOperation(HttpRequestBody requestBody) {
+        if (requestBody.isBinary()) {
```

Comment:
when does this happen?

Reply:
If someone specifies a GraphQL schema in config and happens to hit some POST endpoints which respond with binary data from the same set of tests.

Reply:
But really I was just being defensive.